modules/saas.py

```diff
@@ -0,0 +1,2542 @@
+import sys
+import time
+import warnings
+from pathlib import Path
+
+import requests
+
+warnings.filterwarnings("always", ".*", PendingDeprecationWarning)
+
+
+class HTTPException(Exception):
+    def __init__(self, response):
+        """
+        Custom exception class to report possible API errors
+        The body is contructed by extracting the API error code from the requests.Response object
+        """
+        try:
+            r = response.json()
+            if "detail" in r:
+                detail = r["detail"]
+            elif "errors" in r:
+                detail = r["errors"][0]["title"]
+            elif "_meta" in r:
+                detail = r["_meta"]["message"]
+            else:
+                detail = response.content
+        except Exception:
+            detail = response.content
+        body = f"Status code: {str(response.status_code)} - {detail}"
+        super().__init__(body)
+
+
+class HTTPUnauthorizedException(HTTPException):
+    '''Specific Exception'''
+
+
+class HTTPTooManyRequestsException(HTTPException):
+    '''Specific Exception'''
+
+
+def request_error_handler(func):
+    def request_handler(self, *args, **kwargs):
+        response = func(self, *args, **kwargs)
+        if response.status_code in [200, 201, 204]:
+            return response
+        elif response.status_code == 401:
+            raise HTTPUnauthorizedException(response)
+        elif response.status_code == 429:
+            raise HTTPTooManyRequestsException(response)
+        else:
+            raise HTTPException(response)
+
+    return request_handler
+
+
+def deprecation(message):
+    warnings.warn(message, PendingDeprecationWarning)
+
+
+def param_deprecation(key):
+    message = f"{key} will be deprecated with Vectra API which will be annouced in an upcoming release"
+    warnings.warn(message, PendingDeprecationWarning)
+
+
+class VectraSaaSClient(object):
+    def __init__(self, url=None, client_id=None, secret_key=None, verify=False):
+        """
+        Initialize Vectra Saas client
+        :param url: IP or hostname of Vectra brain (ex https://www.example.com) - required
+        :param client_id: API Client ID for authentication - requried
+        :param secret_key: API Secret Key for authenticaiton - required
+        :param verify: Verify SSL (default: False) - optional
+        """
+        url = VectraSaaSClient._remove_trailing_slashes(url)
+        self.base_url = url
+        self.version = 3
+        self.url = f"{url}/api/v{self.version}"
+        self.verify = verify
+        self._access = None
+
+        self.token_headers = {
+            "Content-Type": "application/x-www-form-urlencoded",
+        }
+
+        if client_id and secret_key:
+            self.auth = (client_id, secret_key)
+        else:
+            raise RuntimeError(
+                "API Client ID and Secret Key are required for authentication."
+            )
+
+    @staticmethod
+    def _remove_trailing_slashes(url):
```

Comment:
This is the same as the method in vectra.py.  It may be worth moving common functions to a shared module.  Also rename the method to better describe what it is now doing.

I am ok if you leave it in vectra.py, and move the method to a module level function that is no longer "private" (prefixed by an underscore). so you could import it from "from vectra import format_url"

Reply:
Sounds good. Any work to remove duplicate code and have it in a shared space makes this easier to maintain long term.